core/src/main/java/io/micronaut/core/naming/NameUtils.java

```diff
@@ -369,7 +369,8 @@ public static boolean isReaderName(@NonNull String methodName, @NonNull String[]
             }
             int len = methodName.length();
             if (len > prefixLength) {
-                isValid = Character.isUpperCase(methodName.charAt(prefixLength));
+                char firstVarNameChar = methodName.charAt(prefixLength);
+                isValid = firstVarNameChar == '_' || Character.isUpperCase(firstVarNameChar);
```

Comment:
I think the correct way would be to use something like `Character.isJavaIdentifierStart`. And the tests should also include a case for `$foobar`

Reply:
I agree about `$` symbol, but how you suggest to use `Character.isJavaIdentifierStart` ? Problem in symbol case `isJavaIdentifierStart` don't check letters case, but symbols `_` and `$` haven't case

Reply:
@dstepanov ping

Reply:
According to JavaDoc: `Determines if the character (Unicode code point) is permissible as the first character in a Java identifier`. Does it pass tests if you replace `isValid = Character.isJavaIdentifierStart(firstVarNameChar)`?

Reply:
```
Determines if the specified character is permissible as the first character in a Java identifier.
A character may start a Java identifier if and only if one of the following conditions is true:
isLetter(ch) returns true
getType(ch) returns LETTER_NUMBER
ch is a currency symbol (such as '$')
ch is a connecting punctuation character (such as '_').
```
It means if you run this code:
```java
        log.info("{}, {}, {}, {}",
                Character.isJavaIdentifierStart('A'),
                Character.isJavaIdentifierStart('a'),
                Character.isJavaIdentifierStart('$'),
                Character.isJavaIdentifierStart('_'));
```
you will see:
```
true, true, true, true
```
In other words, you can't figure out what case a given character is in. Yes, in theory you can capitalize an attribute, like so:
```java
private String Field;
```
but getters and setters must be like this:
```java
    public String getField() {
        return Field;
    }

    public void setField(String field) {
        Field = field;
    }
```
A lowercase letter in a getter can only occur in one case, if you name the variable like this:
```
private String mYfield;
```
In this case, you will get a classic naming problem that many people have - some systems think that after the word get and set there should be an uppercase letter, some think that there should be 2 capital letters. For example, lombok will generate methods like this:
```java
    public String getMYfield() {
        return this.mYfield;
    }

    public void setMYfield(final String mYfield) {
        this.mYfield = mYfield;
    }
```
idea generate it like this:
```java
    public String getmYfield() {
        return mYfield;
    }

    public void setmYfield(String mYfield) {
        this.mYfield = mYfield;
    }
```
How do we view this situation? If we allow that there can be a lowercase letter after the word get and set, then we need to add the condition that after this letter there must be an uppercase letter.

Or do you suggest completely abandoning the character case check and replacing the entire check with one method:
```java
isValid =Character.isJavaIdentifierStart(firstVarNameChar);
```
in this case, absolutely all methods that start with the correct character after get will become valid. For example, method with name `getmyget1223` - will be considered a valid getter. But, as we can see from my examples, that in correct getters and setters there is always at least one uppercase letter, except for the exception with the symbols `_` and `$`:
```java
// this is valid code and right getters and setters

    private String ______;

    public String get______() {
        return ______;
    }

    public void set______(String ______) {
        this.______ = ______;
    }
```

Reply:
It turns out that we can formulate 3 conditions for correctness:

1. Starts with get and contains the characters `_` or `$`
2. Starts with get, does NOT contain the characters `_` or `$` and the next letter is capitalized.
3. Starts with get, does NOT contain the characters `_` or `$` and the next letter in lower case, but the next letter must be in upper case.

Do you agree?

Reply:
Yes, thanks!